File: windows/okfa.windows/Protocol/BridgeProtocol.cs
using System.Buffers.Binary;

namespace KeyboardBridge.Windows.Protocol;

public static class BridgeProtocol
{
    public const byte Version = 1;
}

public static class BridgeUuids
{
    public static readonly Guid Service = Guid.Parse("9F0A1000-6C5B-4E1B-9E15-3A4A7D91C100");
    public static readonly Guid Control = Guid.Parse("9F0A1001-6C5B-4E1B-9E15-3A4A7D91C100");
    public static readonly Guid InputEvent = Guid.Parse("9F0A1002-6C5B-4E1B-9E15-3A4A7D91C100");
    public static readonly Guid InputSnapshot = Guid.Parse("9F0A1003-6C5B-4E1B-9E15-3A4A7D91C100");
}

public enum ControlMessageType : byte
{
    ClientHello = 0x01,
    TrustStatus = 0x02,
    ModeChange = 0x03,
    Heartbeat = 0x04,
    ReleaseAll = 0x05,
    Error = 0x06,
}

public enum TrustStatusCode : byte
{
    Pending = 0,
    Approved = 1,
    Denied = 2,
}

public enum TrustReason : byte
{
    UnknownClient = 1,
    UserApproved = 2,
    UserDenied = 3,
    AutoTrusted = 4,
}

public enum RemoteInputMode : byte
{
    Idle = 0,
    RemoteInputActive = 1,
}

public enum ModeChangeSource : byte
{
    MacUser = 1,
    WindowsUser = 2,
    System = 3,
}

public enum InputEventAction : byte
{
    KeyDown = 1,
    KeyUp = 2,
    KeyRepeat = 3,
    ModifiersOnly = 4,
}

public enum SnapshotReason : byte
{
    ModeStart = 1,
    Periodic = 2,
    ModeStop = 3,
    Resync = 4,
}

public readonly record struct ClientHelloMessage(ushort Sequence, ulong ClientId, ushort ClientVersion, ushort CapabilityFlags);
public readonly record struct TrustStatusMessage(ushort Sequence, TrustStatusCode Status, byte Reason, ushort ServerFlags, ulong ServerId);
public readonly record struct ModeChangeMessage(ushort Sequence, RemoteInputMode Mode, ModeChangeSource Source, ushort Flags);
public readonly record struct ReleaseAllMessage(ushort Sequence);
public readonly record struct InputEventMessage(ushort Sequence, InputEventAction Action, byte HidUsage, byte Modifiers, byte EventFlags, ushort DeltaMilliseconds);
public readonly record struct KeyboardSnapshotMessage(ushort Sequence, SnapshotReason Reason, byte Modifiers, byte PressedCount, IReadOnlyList<byte> Usages);

public static class ControlMessageCodec
{
    public static byte[] EncodeClientHello(ushort sequence, ulong clientId, ushort clientVersion, ushort capabilityFlags)
    {
        var buffer = new byte[16];
        buffer[0] = BridgeProtocol.Version;
        buffer[1] = (byte)ControlMessageType.ClientHello;
        BinaryPrimitives.WriteUInt16LittleEndian(buffer.AsSpan(2, 2), sequence);
        BinaryPrimitives.WriteUInt64LittleEndian(buffer.AsSpan(4, 8), clientId);
        BinaryPrimitives.WriteUInt16LittleEndian(buffer.AsSpan(12, 2), clientVersion);
        BinaryPrimitives.WriteUInt16LittleEndian(buffer.AsSpan(14, 2), capabilityFlags);
        return buffer;
    }

    public static byte[] EncodeTrustStatus(ushort sequence, TrustStatusCode status, byte reason, ushort serverFlags, ulong serverId)
    {
        var buffer = new byte[16];
        buffer[0] = BridgeProtocol.Version;
        buffer[1] = (byte)ControlMessageType.TrustStatus;
        BinaryPrimitives.WriteUInt16LittleEndian(buffer.AsSpan(2, 2), sequence);
        buffer[4] = (byte)status;
        buffer[5] = reason;
        BinaryPrimitives.WriteUInt16LittleEndian(buffer.AsSpan(6, 2), serverFlags);
        BinaryPrimitives.WriteUInt64LittleEndian(buffer.AsSpan(8, 8), serverId);
        return buffer;
    }

    public static byte[] EncodeModeChange(ushort sequence, RemoteInputMode mode, ModeChangeSource source, ushort flags)
    {
        var buffer = new byte[8];
        buffer[0] = BridgeProtocol.Version;
        buffer[1] = (byte)ControlMessageType.ModeChange;
        BinaryPrimitives.WriteUInt16LittleEndian(buffer.AsSpan(2, 2), sequence);
        buffer[4] = (byte)mode;
        buffer[5] = (byte)source;
        BinaryPrimitives.WriteUInt16LittleEndian(buffer.AsSpan(6, 2), flags);
        return buffer;
    }

    public static byte[] EncodeReleaseAll(ushort sequence)
    {
        var buffer = new byte[4];
        buffer[0] = BridgeProtocol.Version;
        buffer[1] = (byte)ControlMessageType.ReleaseAll;
        BinaryPrimitives.WriteUInt16LittleEndian(buffer.AsSpan(2, 2), sequence);
        return buffer;
    }

    public static bool TryDecodeClientHello(byte[] payload, out ClientHelloMessage message)
    {
        message = default;

        if (payload.Length < 16)
        {
            return false;
        }

        if (payload[0] != BridgeProtocol.Version || payload[1] != (byte)ControlMessageType.ClientHello)
        {
            return false;
        }

        message = new ClientHelloMessage(
            Sequence: BinaryPrimitives.ReadUInt16LittleEndian(payload.AsSpan(2, 2)),
            ClientId: BinaryPrimitives.ReadUInt64LittleEndian(payload.AsSpan(4, 8)),
            ClientVersion: BinaryPrimitives.ReadUInt16LittleEndian(payload.AsSpan(12, 2)),
            CapabilityFlags: BinaryPrimitives.ReadUInt16LittleEndian(payload.AsSpan(14, 2))
        );

        return true;
    }

    public static bool TryDecodeTrustStatus(byte[] payload, out TrustStatusMessage message)
    {
        message = default;

        if (payload.Length < 16)
        {
            return false;
        }

        if (payload[0] != BridgeProtocol.Version || payload[1] != (byte)ControlMessageType.TrustStatus)
        {
            return false;
        }

        message = new TrustStatusMessage(
            Sequence: BinaryPrimitives.ReadUInt16LittleEndian(payload.AsSpan(2, 2)),
            Status: (TrustStatusCode)payload[4],
            Reason: payload[5],
            ServerFlags: BinaryPrimitives.ReadUInt16LittleEndian(payload.AsSpan(6, 2)),
            ServerId: BinaryPrimitives.ReadUInt64LittleEndian(payload.AsSpan(8, 8))
        );

        return true;
    }

    public static bool TryDecodeModeChange(byte[] payload, out ModeChangeMessage message)
    {
        message = default;

        if (payload.Length < 8)
        {
            return false;
        }

        if (payload[0] != BridgeProtocol.Version || payload[1] != (byte)ControlMessageType.ModeChange)
        {
            return false;
        }

        message = new ModeChangeMessage(
            Sequence: BinaryPrimitives.ReadUInt16LittleEndian(payload.AsSpan(2, 2)),
            Mode: (RemoteInputMode)payload[4],
            Source: (ModeChangeSource)payload[5],
            Flags: BinaryPrimitives.ReadUInt16LittleEndian(payload.AsSpan(6, 2))
        );

        return true;
    }

    public static bool TryDecodeReleaseAll(byte[] payload, out ReleaseAllMessage message)
    {
        message = default;

        if (payload.Length < 4)
        {
            return false;
        }

        if (payload[0] != BridgeProtocol.Version || payload[1] != (byte)ControlMessageType.ReleaseAll)
        {
            return false;
        }

        message = new ReleaseAllMessage(
            Sequence: BinaryPrimitives.ReadUInt16LittleEndian(payload.AsSpan(2, 2))
        );

        return true;
    }
}

public static class InputMessageCodec
{
    public static byte[] EncodeInputEvent(InputEventMessage message)
    {
        var buffer = new byte[12];
        buffer[0] = BridgeProtocol.Version;
        buffer[1] = 0x10;
        BinaryPrimitives.WriteUInt16LittleEndian(buffer.AsSpan(2, 2), message.Sequence);
        buffer[4] = (byte)message.Action;
        buffer[5] = message.HidUsage;
        buffer[6] = message.Modifiers;
        buffer[7] = message.EventFlags;
        BinaryPrimitives.WriteUInt16LittleEndian(buffer.AsSpan(8, 2), message.DeltaMilliseconds);
        BinaryPrimitives.WriteUInt16LittleEndian(buffer.AsSpan(10, 2), 0);
        return buffer;
    }

    public static byte[] EncodeSnapshot(KeyboardSnapshotMessage message)
    {
        var buffer = new byte[14];
        buffer[0] = BridgeProtocol.Version;
        buffer[1] = 0x20;
        BinaryPrimitives.WriteUInt16LittleEndian(buffer.AsSpan(2, 2), message.Sequence);
        buffer[4] = (byte)message.Reason;
        buffer[5] = message.Modifiers;

        var clippedUsages = message.Usages.Take(6).ToArray();
        buffer[6] = (byte)Math.Min(message.PressedCount, clippedUsages.Length);
        buffer[7] = 0;
        for (var index = 0; index < clippedUsages.Length; index += 1)
        {
            buffer[8 + index] = clippedUsages[index];
        }

        return buffer;
    }

    public static bool TryDecodeInputEvent(byte[] payload, out InputEventMessage message)
    {
        message = default;

        if (payload.Length < 12)
        {
            return false;
        }

        if (payload[0] != BridgeProtocol.Version || payload[1] != 0x10)
        {
            return false;
        }

        message = new InputEventMessage(
            Sequence: BinaryPrimitives.ReadUInt16LittleEndian(payload.AsSpan(2, 2)),
            Action: (InputEventAction)payload[4],
            HidUsage: payload[5],
            Modifiers: payload[6],
            EventFlags: payload[7],
            DeltaMilliseconds: BinaryPrimitives.ReadUInt16LittleEndian(payload.AsSpan(8, 2))
        );

        return true;
    }

    public static bool TryDecodeSnapshot(byte[] payload, out KeyboardSnapshotMessage message)
    {
        message = default;

        if (payload.Length < 14)
        {
            return false;
        }

        if (payload[0] != BridgeProtocol.Version || payload[1] != 0x20)
        {
            return false;
        }

        var pressedCount = payload[6];
        var rawUsages = payload.Skip(8).Take(6).Where(static usage => usage != 0).ToArray();
        var clippedCount = (byte)Math.Min(pressedCount, rawUsages.Length);
        var usages = rawUsages.Take(clippedCount).ToArray();

        message = new KeyboardSnapshotMessage(
            Sequence: BinaryPrimitives.ReadUInt16LittleEndian(payload.AsSpan(2, 2)),
            Reason: (SnapshotReason)payload[4],
            Modifiers: payload[5],
            PressedCount: clippedCount,
            Usages: usages
        );

        return true;
    }
}

File: mac/Sources/Protocol/BridgeProtocol.swift
import CoreBluetooth
import Foundation

enum BridgeProtocol {
    static let version: UInt8 = 1
}

enum BridgeUUIDs {
    static let service = CBUUID(string: "9F0A1000-6C5B-4E1B-9E15-3A4A7D91C100")
    static let control = CBUUID(string: "9F0A1001-6C5B-4E1B-9E15-3A4A7D91C100")
    static let inputEvent = CBUUID(string: "9F0A1002-6C5B-4E1B-9E15-3A4A7D91C100")
    static let inputSnapshot = CBUUID(string: "9F0A1003-6C5B-4E1B-9E15-3A4A7D91C100")
}

enum ControlMessageType: UInt8 {
    case clientHello = 0x01
    case trustStatus = 0x02
    case modeChange = 0x03
    case heartbeat = 0x04
    case releaseAll = 0x05
    case error = 0x06
}

enum TrustStatusCode: UInt8 {
    case pending = 0
    case approved = 1
    case denied = 2
}

enum TrustReason: UInt8 {
    case unknownClient = 1
    case userApproved = 2
    case userDenied = 3
    case autoTrusted = 4
}

enum RemoteInputMode: UInt8 {
    case idle = 0
    case remoteInputActive = 1
}

enum ModeChangeSource: UInt8 {
    case macUser = 1
    case windowsUser = 2
    case system = 3
}

enum InputEventAction: UInt8 {
    case keyDown = 1
    case keyUp = 2
    case keyRepeat = 3
    case modifiersOnly = 4
}

enum SnapshotReason: UInt8 {
    case modeStart = 1
    case periodic = 2
    case modeStop = 3
    case resync = 4
}

struct ControlHeader {
    let version: UInt8
    let type: UInt8
    let sequence: UInt16
}

struct ClientHelloMessage {
    let header: ControlHeader
    let clientId: UInt64
    let clientVersion: UInt16
    let capabilityFlags: UInt16
}

struct TrustStatusMessage {
    let header: ControlHeader
    let status: TrustStatusCode
    let reason: UInt8
    let serverFlags: UInt16
    let serverId: UInt64
}

struct ModeChangeMessage {
    let header: ControlHeader
    let mode: RemoteInputMode
    let source: ModeChangeSource
    let flags: UInt16
}

struct KeyEventMessage {
    let sequence: UInt16
    let action: InputEventAction
    let hidUsage: UInt8
    let modifiers: UInt8
    let eventFlags: UInt8
    let deltaMilliseconds: UInt16

    func encode() -> Data {
        var data = Data()
        data.append(BridgeProtocol.version)
        data.append(0x10)
        data.appendLittleEndian(sequence)
        data.append(action.rawValue)
        data.append(hidUsage)
        data.append(modifiers)
        data.append(eventFlags)
        data.appendLittleEndian(deltaMilliseconds)
        data.appendLittleEndian(UInt16(0))
        return data
    }
}

struct KeyboardSnapshotMessage {
    let sequence: UInt16
    let reason: SnapshotReason
    let modifiers: UInt8
    let usages: [UInt8]

    func encode() -> Data {
        var data = Data()
        data.append(BridgeProtocol.version)
        data.append(0x20)
        data.appendLittleEndian(sequence)
        data.append(reason.rawValue)
        data.append(modifiers)
        data.append(UInt8(usages.count))
        data.append(0)

        let clipped = Array(usages.prefix(6))
        for usage in clipped {
            data.append(usage)
        }
        if clipped.count < 6 {
            data.append(Data(repeating: 0, count: 6 - clipped.count))
        }

        return data
    }
}

enum ControlMessageCodec {
    static func decodeClientHello(_ data: Data) -> ClientHelloMessage? {
        guard data.count >= 16 else {
            return nil
        }

        let version = data[0]
        let type = data[1]
        guard version == BridgeProtocol.version, type == ControlMessageType.clientHello.rawValue else {
            return nil
        }

        let header = ControlHeader(
            version: version,
            type: type,
            sequence: data.readUInt16LE(at: 2)
        )

        return ClientHelloMessage(
            header: header,
            clientId: data.readUInt64LE(at: 4),
            clientVersion: data.readUInt16LE(at: 12),
            capabilityFlags: data.readUInt16LE(at: 14)
        )
    }

    static func encodeTrustStatus(
        sequence: UInt16,
        status: TrustStatusCode,
        reason: TrustReason,
        serverFlags: UInt16,
        serverId: UInt64
    ) -> Data {
        var data = Data()
        data.append(BridgeProtocol.version)
        data.append(ControlMessageType.trustStatus.rawValue)
        data.appendLittleEndian(sequence)
        data.append(status.rawValue)
        data.append(reason.rawValue)
        data.appendLittleEndian(serverFlags)
        data.appendLittleEndian(serverId)
        return data
    }

    static func encodeModeChange(sequence: UInt16, mode: RemoteInputMode, source: ModeChangeSource, flags: UInt16) -> Data {
        var data = Data()
        data.append(BridgeProtocol.version)
        data.append(ControlMessageType.modeChange.rawValue)
        data.appendLittleEndian(sequence)
        data.append(mode.rawValue)
        data.append(source.rawValue)
        data.appendLittleEndian(flags)
        return data
    }

    static func encodeReleaseAll(sequence: UInt16) -> Data {
        var data = Data()
        data.append(BridgeProtocol.version)
        data.append(ControlMessageType.releaseAll.rawValue)
        data.appendLittleEndian(sequence)
        return data
    }
}

extension Data {
    mutating func appendLittleEndian(_ value: UInt16) {
        var littleEndianValue = value.littleEndian
        Swift.withUnsafeBytes(of: &littleEndianValue) { buffer in
            append(buffer.bindMemory(to: UInt8.self))
        }
    }

    mutating func appendLittleEndian(_ value: UInt64) {
        var littleEndianValue = value.littleEndian
        Swift.withUnsafeBytes(of: &littleEndianValue) { buffer in
            append(buffer.bindMemory(to: UInt8.self))
        }
    }

    func readUInt16LE(at offset: Int) -> UInt16 {
        let end = offset + MemoryLayout<UInt16>.size
        let range = offset..<end
        return subdata(in: range).withUnsafeBytes { $0.load(as: UInt16.self) }.littleEndian
    }

    func readUInt64LE(at offset: Int) -> UInt64 {
        let end = offset + MemoryLayout<UInt64>.size
        let range = offset..<end
        return subdata(in: range).withUnsafeBytes { $0.load(as: UInt64.self) }.littleEndian
    }

    var hexString: String {
        map { String(format: "%02X", $0) }.joined(separator: " ")
    }
}

extension UInt64 {
    var hexClientId: String {
        String(format: "%016llX", self)
    }
}
